internal/pkg/agent/application/upgrade/watcher.go

```diff
@@ -0,0 +1,217 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package upgrade
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/elastic/elastic-agent/pkg/control/v2/client"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+const (
+	statusCheckMissesAllowed      = 4 // enable 2 minute start (30 second periods)
+	statusLossesAllowed           = 2 // enable connection lost to agent twice
+	statusFailureFlipFlopsAllowed = 3 // no more than three failure flip-flops allowed
+)
+
+var (
+	// ErrCannotConnect is returned when connection cannot be made to the agent.
+	ErrCannotConnect = errors.New("failed to connect to agent daemon")
+	// ErrLostConnection is returned when connection is lost to the agent daemon.
+	ErrLostConnection = errors.New("lost connection to agent daemon")
+	// ErrAgentStatusFailed is returned when agent reports FAILED status.
+	ErrAgentStatusFailed = errors.New("agent reported failed state")
+	// ErrAgentComponentFailed is returned when agent reports FAILED status for a component
+	ErrAgentComponentFailed = errors.New("agent reported failed component(s) state")
+	// ErrAgentFlipFlopFailed is returned when agent flip-flops between failed and healthy.
+	ErrAgentFlipFlopFailed = errors.New("agent reported on and off failures ")
+)
+
+// AgentWatcher watches for the ability to connect to the running Elastic Agent, if it reports any errors
+// and how many times it disconnects from the Elastic Agent while running.
+type AgentWatcher struct {
+	connectCounter int
+	lostCounter    int
+
+	notifyChan    chan error
+	log           *logger.Logger
+	agentClient   client.Client
+	checkInterval time.Duration
+}
+
+// NewAgentWatcher creates a new agent watcher.
+func NewAgentWatcher(ch chan error, log *logger.Logger, checkInterval time.Duration) *AgentWatcher {
+	c := client.New()
+	ec := &AgentWatcher{
+		notifyChan:    ch,
+		agentClient:   c,
+		log:           log,
+		checkInterval: checkInterval,
+	}
+	return ec
+}
+
+// Run runs the checking loop.
+func (ch *AgentWatcher) Run(ctx context.Context) {
+	ch.log.Info("Agent watcher started")
+
+	ch.connectCounter = 0
+	ch.lostCounter = 0
+
+	// tracking of an error runs in a separate goroutine, because
+	// the call to `watch.Recv` blocks and a timer is needed
+	// to determine if an error last longer than the checkInterval.
+	failedReset := make(chan bool)
+	failedCh := make(chan error)
+	go func() {
+		failedTimer := time.NewTimer(ch.checkInterval)
+		failedTimer.Stop()       // starts stopped
+		defer failedTimer.Stop() // stopped on exit always
+
+		var failedCount int
+		var failedErr error
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case reset := <-failedReset:
+				if reset {
+					failedCount = 0
+					failedTimer.Stop()
+				}
+			case err := <-failedCh:
+				if err != nil {
+					if failedErr == nil {
+						failedCount++
```

Comment:
It looks like we only increment `failedCount` when we received a non-nil error from `failedCh` but previously there was no error (i.e. `failedErr == nil`).  So that means `failedCount` actually represents the number of times we flip-flopped from a non-error state to an error state, right?  If my understanding is correct, could we rename this to something like `flipFlopCount` or something else indicating that it's counting flip-flops?

Reply:
Done.